src/distance.cpp

```diff
@@ -0,0 +1,620 @@
+         first node */
+       handle_t handle = graph->get_handle(start.first, start.second);
+       vector<const handle_t*> nextNodes;//vector of next nodes
+       auto addHandle = [&](const handle_t& h)-> bool {
+           nextNodes.push_back(&h);
+           return true;
+       };
+       graph->follow_edges(handle, false, addHandle);
+       handle_t nextHandle = *nextNodes[0];
+       size_t j = index(start, make_pair(graph->get_id(nextHandle), 
+                                         graph->get_is_reverse(nextHandle)));
+
+       int64_t firstDist = distances->at(j);//Length of first node 
+       return totalDist - firstDist;
+    }
+}
+
+ChainDistances::ChainDistances(unordered_map<id_t, size_t>* s, 
+                                                          vector<int64_t>* p) {
+    
+    snarl_to_index = s;
+    prefix_sum = p;
+}
+
+int64_t ChainDistances::chainDistance(pair<id_t, bool> start, 
+                                                      pair<id_t, bool> end) {
+    //Distance between start of start and start of end in a chain
+    size_t i1 = snarl_to_index->at(start.first);
+    size_t i2 = snarl_to_index->at(end.first); 
+    if ((start.second && end.second) || (!start.second && !end.second)) {
+        //If start and end are facing the same direction
+        int64_t dNoRev = abs(prefix_sum->at(i1) - prefix_sum->at(i2)); 
+        return dNoRev;
+    } else {
+        //TODO Currently assumes no reversing edges in the chain
+        return -1;
+    }
+}
+
+int64_t ChainDistances::chainDistance_short(VG* graph, pair<id_t, bool> start,
+                                                      pair<id_t, bool> end) {
+    //Distance between end of start node to beginning of end node in chain
+    
+    size_t i1 = snarl_to_index->at(start.first);
+    size_t i2 = snarl_to_index->at(end.first); 
+    int64_t d1 = chainDistance(start, end);
+    int64_t d2 = chainDistance(end, start);
+    if (d1 == -1 && d2 == -1) {
+        return -1;
+    } else if (d1 == -1) {
+        return d2 - graph->get_node(i1)->sequence().size();
+    } else if (d2 == -1) {
+        return d1 - graph->get_node(i2)->sequence().size();
+    } else {
+        return min(d2 - graph->get_node(i1)->sequence().size(), 
+                   d1 - graph->get_node(i2)->sequence().size());
+    }
+}
+
+int64_t ChainDistances::chainLength() {
+    //Get the length of a chain including length of last node
+    return prefix_sum->back();
+}
+
+
+int64_t calculateIndex(DistanceIndex* di, VG* graph, SnarlManager* sm, 
+         const Chain* chain) {
+    /*Add chain to distance index and recursively add all distances in 
+      snarls contained within chain
+    */
+    auto cmp = [] (pair<const handle_t*,int64_t> x,
+                                           pair<const handle_t*,int64_t> y) {
+        //Comparison function for the priority of a pair of handle, distance
+        return (x.second < y.second); 
```

Comment:
I believe `priority_queue`s are designed to return the max value, so you want to have it return the min distance (to do a Dijkstra traversal) I think you'll need to reverse the order of this comparator.